STM32: Add support for STM32H7 dual core

erwango opened the pull request:
Enable support of Zephyr on STM32H747I-DISCO board.
This board is dual core and this work enable running zephyr in both cores.

For now, the minimum work is done to enable dual core operation by sharing the resources.
No communication is enabled between cores except use of hardware semaphore for synchronization
and critical resources access.

Depends on https://github.com/zephyrproject-rtos/hal_stm32/pull/15

erwango committed:
drivers/pinmux: stm32: Add UART headers for STM32H7

Add UART headers for STM32H7 series

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
dts: stm32h7: Add dtsi for m4 core

This change adds sram1 plus m4 core package level dtsi.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
drivers/clock_control: stm32h7: Disable configuration for CM4 core

On STM32H7, in Dual core configuration, we restrict configuration
access to CM7 core. CM4 can access to API but not the init part of
the driver.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
west: Bump hal_stm32 to version supporting STM32H7 series

Now STM32H7 are available in Zephyr.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
dts: STM32H747 basic device tree description

Initiate stm32h7 device tree description, with stm32h747 single core
configuration.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
modules: update stm32cube Kconfig with H7

Following introduction of STM32H7 series, update STM32 module
Kconfig with stm32cube H7 additions.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango commented:
@arnop2

erwango committed:
drivers: stm32: Fix leading space warnings

In two stm32 drivers, fix leading spaces warning.


Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango commented:
`stm32h747i_disco_m4`, doesn't enable `CONFIG_SERIAL`.

How come `samples/nfc/nfc_hello` (filters on `CONFIG_SERIAL`) and `samples/mpu/mpu_test` (filters on `CONFIG_SERIAL_SUPPORT_INTERRUPT`) are compiled on that target ?

dbkinder commented on boards/arm/stm32h747i_disco/doc/index.rst:
```diff
@@ -0,0 +1,209 @@
+
+The STM32H747I Discovery kit has up to 8 UARTs.
+Default configuration assigns USART1 and UART8 to the CPU1. The Zephyr console
+output is assigned to UART1 which connected to the onboard ST-LINK/V. Virtual
```
Not sure what the right fix is here, but should this be ``ST-LINK/Virtual``

erwango replied:
Should be ST-LINK/V3.0. Txs

dbkinder commented on boards/arm/stm32h747i_disco/doc/index.rst:
```diff
@@ -0,0 +1,209 @@
+- Compilation: Clock configuration is only accessible to M7 core. M4 core only
+  have access to bus clock activation, deactivation
+- Static pre-compilation assignment: Peripherals such as UART are assigned in device
+  tree before compilation. it is up to the user to ensure same peripheral is not
```
```suggestion
  tree before compilation. The user must ensure peripherals are not
```

erwango replied:
Done

dbkinder commented on boards/arm/stm32h747i_disco/doc/index.rst:
```diff
@@ -0,0 +1,209 @@
+  have access to bus clock activation, deactivation
+- Static pre-compilation assignment: Peripherals such as UART are assigned in device
+  tree before compilation. it is up to the user to ensure same peripheral is not
+  assigned to both cores at the same time
```
```suggestion
  assigned to both cores at the same time.
```

erwango replied:
Done

dbkinder commented on boards/arm/stm32h747i_disco/doc/index.rst:
```diff
@@ -0,0 +1,209 @@
+option bytes configuration.
+It is advised to use `STM32CubeProgrammer`_ to check and update option bytes
+configuration and flash ``stm32h747i_disco_m7`` and ``stm32h747i_disco_m4`` targets.
+By default:
```
I'd start a new paragraph here by adding a blank line before ``By default:``

erwango replied:
Done

dbkinder reviewed (changes requested):
some recommended doc changes

erwango commented on boards/arm/stm32h747i_disco/doc/index.rst:
```diff
@@ -0,0 +1,209 @@
+
+The discovery kit enables a wide diversity of applications taking benefit
+from audio, multi-sensor support, graphics, security, security, video,
+and high-speed connectivity features. Important board features include:
```
Done

erwango commented on boards/arm/stm32h747i_disco/doc/index.rst:
```diff
@@ -0,0 +1,209 @@
+from audio, multi-sensor support, graphics, security, security, video,
+and high-speed connectivity features. Important board features include:
+
+It embeds STM32H747XI SoC: a high-performance and DSP with DP-FPU, Arm Cortex-M7 + Cortex-M4 MCU,
```
Done

erwango commented on boards/arm/stm32h747i_disco/doc/index.rst:
```diff
@@ -0,0 +1,209 @@
+
+It embeds STM32H747XI SoC: a high-performance and DSP with DP-FPU, Arm Cortex-M7 + Cortex-M4 MCU,
+with 2MBytes of Flash memory, 1MB RAM, 480 MHz CPU, Art Accelerator, L1 cache, external memory interface,
+large set of peripherals, SMPS, MIPI-DSI.
```
Done

erwango commented on boards/arm/stm32h747i_disco/doc/index.rst:
```diff
@@ -0,0 +1,209 @@
+Other hardware features are not yet supported on Zephyr porting.
+
+The default configuration per core can be found in the defconfig files:
+``boards/arm/stm32h747i_disco/stm32h747i_disco_defconfig_m7``
```
Done

arnop2 commented on boards/arm/stm32h747i_disco/Kconfig.defconfig:
```diff
@@ -0,0 +1,27 @@
+if UART_CONSOLE
+
+config UART_1
+	default y if BOARD_STM32H747I_DISCO_M7
```
As peripherals are shared between cortex contexts i would set configs to n by default for peripheral to ensure that is not used by M4 and M7

erwango replied:
Problem is that I need one UART defined on one side to enable serial testing (and things that require serial).

But, I agree with your concern and this is actually why I preferred to put devices activation in this file common to both cores rather than in _m{4.7}_defconfig files.
What I can add is the following to make the intention clearer (won't add nothing in term of "mechanical" check though :
```
config UART_1
	default y if BOARD_STM32H747I_DISCO_M7
	default n if BOARD_STM32H747I_DISCO_M4
```
And the same should be done for each other peripheral device.

arnop2 replied:
ok for me

arnop2 commented on boards/arm/stm32h747i_disco/stm32h747i_disco.dtsi:
```diff
@@ -0,0 +1,57 @@
+		};
+	};
+};
+
```
Which core  is in charge of these IO, the M4 or the M7? i would declare it in related dtsi...
seem(IMHO) that this file should declare only peripheral shared by both the M4 and the M7 ( probably protected by HSEM protection).

erwango replied:
Well, this is theoretically accessible to both cores, so I found it made sense to define in a common file.
Though please note they should theoretically be accessed only though the aliases, which are split per core .dts

arnop2 replied:
This new implementation sounds better for me. stm32h747i_disco.dtsi declares resources available on boards,   stm32h747i_disco_m7.dts and stm32h747i_disco_m4.dts enable them for use.
 you can mark it as resolved

erwango committed:
soc/arm: Provide basic support for STM32H7 series

Enable basic support to STM32H7, in single core configuration (M7).

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
drivers/clock_control: STM32H7 support

Provide basic clock control driver for STM32H7.
Bus clock activation is done through CM7 and CM4 common registers
so we don't have to care to the CPU Id before accessing.
Accesses are not protected for now. Only possible configuration
is system clock source set to HSE driven PLL.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
drivers/interrupt_controller: stm32: Add support for stm32h7

Add bare minimum to enable EXTI on STM32H7,
in single core configuration.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
drivers/gpio: Add support on STM32H7 series

Add GPIO support on STM32H7.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
drivers/serial: stm32: Add support for STM32H7 series

Add needful to enable uart on STM32H7.
This mostly impact dts but as well soc for fixup.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
soc: stm32h7: Add support for Cortex-M4 core

Add support for C-M4 core on STM32H7 series.
It is enabled in Dual core context with 2 alternatives boot methods:
* Boot CM4 CM7: Both core boot at reset, then CM4 enters Stop mode.
CM7 performs system configuration then finally wakes up CM4
* Boot CM7, CM4 Gated: Only CM7 boots at reset. Once done with
system configuration it triggers (requires option byte update)

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
drivers/gpio: stm32: Add semaphore on gpio_configure

Protect gpio_configure function in dual core context.
This operation is not needed for other fuctions of the api:
* init
* read
* write

Protecting gpio_configure also protects access to
interrupt_controller IP.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango commented on soc/arm/st_stm32/stm32h7/soc_m4.c:
```diff
@@ -0,0 +1,92 @@
+	/*HW semaphore Clock enable*/
+	LL_AHB4_GRP1_EnableClock(LL_AHB4_GRP1_PERIPH_HSEM);
+
+#if defined(CONFIG_STM32H7_BOOT_CM4_CM7)
```
This was my first try, but then, I need to define `stm32h7_m4_boot_stop` also when `CONFIG_STM32H7_BOOT_CM4_CM7` is not enabled, so I preferred to keep it this way

arnop2 replied:
ok  for me

erwango commented on drivers/clock_control/clock_stm32_ll_h7.c:
```diff
@@ -32,6 +32,7 @@
 /**
  * @brief fill in AHB/APB buses configuration structure
  */
+#if !defined(CONFIG_CPU_CORTEX_M4)
```
As discussed, I find it clearer this way. But thanks for proposal!

erwango commented on drivers/clock_control/clock_stm32_ll_h7.c:
```diff
@@ -197,6 +205,8 @@ static int stm32_clock_control_init(struct device *dev)
 {
 	ARG_UNUSED(dev);
 
+#if !defined(CONFIG_CPU_CORTEX_M4)
```
Same as previous, would require `config_bus_prescalers` being defined in both cases, and then getting warning because not used..

erwango committed:
boards: Add stm32h747i_disco, C-M7 target

Add C-M7 target for board stm32h747i_disco.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
boards: stm32h747i_disco: Add m4 target

Add m4 target to stm32h747i_disco.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

dbkinder commented on boards/arm/stm32h747i_disco/doc/index.rst:
```diff
@@ -0,0 +1,210 @@
+********
+
+The discovery kit enables a wide diversity of applications taking benefit
+from audio, multi-sensor support, graphics, security, security, video,
```
```suggestion
from audio, multi-sensor support, graphics, security, video,
```

erwango replied:
done

dbkinder commented on boards/arm/stm32h747i_disco/doc/index.rst:
```diff
@@ -0,0 +1,210 @@
+Resources sharing
+=================
+
+The dual core nature of STM32H747 SoC imposes to share HW resources between the
```
```suggestion
The dual core nature of STM32H747 SoC requires sharing HW resources between the
```

erwango replied:
done

dbkinder commented on boards/arm/stm32h747i_disco/doc/index.rst:
```diff
@@ -0,0 +1,210 @@
+The dual core nature of STM32H747 SoC imposes to share HW resources between the
+two cores. This is done in 3 ways:
+
+- Compilation: Clock configuration is only accessible to M7 core. M4 core only
```
```suggestion
- **Compilation**: Clock configuration is only accessible to the M7 core. The M4 core only
```

erwango replied:
Done

dbkinder commented on boards/arm/stm32h747i_disco/doc/index.rst:
```diff
@@ -0,0 +1,210 @@
+two cores. This is done in 3 ways:
+
+- Compilation: Clock configuration is only accessible to M7 core. M4 core only
+  have access to bus clock activation, deactivation
```
```suggestion
  has access to bus clock activation and deactivation.
```

erwango replied:
Done

dbkinder commented on boards/arm/stm32h747i_disco/doc/index.rst:
```diff
@@ -0,0 +1,210 @@
+
+- Compilation: Clock configuration is only accessible to M7 core. M4 core only
+  have access to bus clock activation, deactivation
+- Static pre-compilation assignment: Peripherals such as UART are assigned in device
```
```suggestion
- **Static pre-compilation assignment**: Peripherals such as a UART are assigned in device
```

erwango replied:
Done

dbkinder commented on boards/arm/stm32h747i_disco/doc/index.rst:
```diff
@@ -0,0 +1,210 @@
+- Static pre-compilation assignment: Peripherals such as UART are assigned in device
+  tree before compilation. The user must ensure peripherals are not assigned to
+  both cores at the same time.
+- Run time protection: Interrupt-controller and GPIO configuration could be accessed
```
```suggestion
- **Run time protection**: Interrupt-controller and GPIO configurations could be accessed
```

erwango replied:
Done

dbkinder commented on boards/arm/stm32h747i_disco/doc/index.rst:
```diff
@@ -0,0 +1,210 @@
+  tree before compilation. The user must ensure peripherals are not assigned to
+  both cores at the same time.
+- Run time protection: Interrupt-controller and GPIO configuration could be accessed
+  by both core at run time. Accesses are protected by hardware semaphore to avoid
```
```suggestion
  by both cores at run time. Accesses are protected by a hardware semaphore to avoid
```

erwango replied:
Done

dbkinder commented on boards/arm/stm32h747i_disco/doc/index.rst:
```diff
@@ -0,0 +1,210 @@
+*************************
+
+Applications for the ``stm32h747i_disco`` board should be built per core target,
+using either stm32h747i_disco_m7 or stm32h747i_disco_m4as targets.
```
```suggestion
using either ``stm32h747i_disco_m7`` or ``stm32h747i_disco_m4`` as the target.
```

erwango replied:
Done

dbkinder commented on boards/arm/stm32h747i_disco/doc/index.rst:
```diff
@@ -0,0 +1,210 @@
+
+Applications for the ``stm32h747i_disco`` board should be built per core target,
+using either stm32h747i_disco_m7 or stm32h747i_disco_m4as targets.
+See :ref:`build_an_application` for more information about applications build.
```
```suggestion
See :ref:`build_an_application` for more information about application builds.
```

erwango replied:
Done

dbkinder commented on boards/arm/stm32h747i_disco/doc/index.rst:
```diff
@@ -0,0 +1,210 @@
+   :board: stm32h747i_disco_m7
+   :goals: build
+
+Use following commands to flash either m7 or m4 target:
```
```suggestion
Use the following commands to flash either the m7 or m4 target:
```

erwango replied:
Done

erwango committed:
boards: stm32h747i_disco: Provide doc

Provide doc for stm32h747i_disco.
Includes basic description for building and flashing
individual cores.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

galak commented on dts/arm/st/h7/stm32h7.dtsi:
```diff
@@ -0,0 +1,57 @@
+			mpu: mpu@e000ed90 {
+				compatible = "arm,armv7m-mpu";
+				reg = <0xe000ed90 0x40>;
+				arm,num-mpu-regions = <8>;
```
Any reason the MPU is in the CPU nodes?

erwango replied:
I actually followed the scheme provided for STM32F7.
It's a recent addtion from @ioannisg : https://github.com/zephyrproject-rtos/zephyr/pull/12173

ioannisg replied:
Is this Cortex-M3/4 @erwango ? then you don't really need that; it's going to be 8 for sure.
You need this for CM7 or CM33

erwango replied:
@ioannisg, this one is CM7.

dbkinder reviewed (approved):
doc changes LGTM, thanks!

ioannisg commented on boards/arm/stm32h747i_disco/Kconfig.board:
```diff
@@ -0,0 +1,16 @@
+config BOARD_STM32H747I_DISCO_M7
+	bool "STM32H747I Discovery Development Board"
+	depends on SOC_STM32H747XX
+	select CPU_CORTEX_M7
```
Weird that Board options select Cortex cores. 
could this be done in the SOC level instead?

erwango replied:
This was my initial attempt, but at the end, this only introduces and intermediate SOC_STM32H747XX_M7, that selects CPU_CORTEX_M7.
When electing the cortex at this level, I can remove an intermediate step which makes things clearer. Still, we're able to do cortex specific settings in soc/arm/st_stm32/stm32h7/Kconfig.soc

ioannisg commented on boards/arm/stm32h747i_disco/Kconfig.board:
```diff
@@ -0,0 +1,16 @@
+config BOARD_STM32H747I_DISCO_M4
+	bool "STM32H747I Discovery Development Board"
+	depends on SOC_STM32H747XX
+	select CPU_CORTEX_M4
```
same here

ioannisg reviewed (commented):
Looks good, I only notice that CPU core options are selected at the board level.
That's not what we have done in the other multi-core SOCs (but I might be wrong).

I think it would be cleaner, if at the SOC level we have SOC options for the two cores, i.e. xxx_M4 and xxx_M7 and then we have two different BOARDs, anyways, because we need to build two different images.

erwango commented:
@ioannisg 

> I think it would be cleaner, if at the SOC level we have SOC options for the two cores, i.e. xxx_M4 and xxx_M7

As answered above, this would introduce an intermediate symbol with no added value.
And I don't see limitation or risk with the current method

> we have two different BOARDs, anyways, because we need to build two different images. 

I understand you propose to introduce 2 boards directories.
I don't think this is the good direction: even if we have 2 targets to build, they share a lot of common configuration: boot type, clocks settings, boards support package, pinmux ..
Having 2 targets in the same board directory allow to share some Kconfig file which force configuration coherency.
Also, even if we miss a good tool to handle peripheral assignment, having all information at one place eases manual check that resources are assigned to only one core at the same time.

ioannisg commented:
@erwango, got me a bit wrong, here, we shoudln't have 2 BOARD directories, we should have two BOARD .defconfigs (we do actually), as we do for other dual-cores.

So, I am only worried about consistency here. Check the 2 dual-core NXP lpc platforms; they have two SOC symbols defined, for each core, and the two BOARD symbols depend on either SOC Kconfig. I think the ARM Musca_x does the same. We should have one way to do this for any multi-core platform....

erwango commented:
@ioannisg 

> @erwango, got me a bit wrong, here, we shoudln't have 2 BOARD directories, we should have two BOARD .defconfigs (we do actually), as we do for other dual-cores.

Ok. And we have 2 _defconfig, but only one .defconfig. This is the same than lpx as I can check. So we're aligned here.

> So, I am only worried about consistency here. Check the 2 dual-core NXP lpc platforms; they have two SOC symbols defined, for each core, and the two BOARD symbols depend on either SOC Kconfig. I think the ARM Musca_x does the same. We should have one way to do this for any multi-core platform.

From what I see, ARM musca defines 2 boards, so does cypress_062_wifi_bt dual core board, let put these aside. 
I see your point about consistency. But then, looking at lpc5Xxxx implementations, I don't particularly agree on the way things are done (specially in soc files Kconfig.defconfig.lpcxxxx_mx some kind of peripheral assignment is done, which limits the flexibility at board level and performs driver deactivation at soc level which seems error prone for an end user).
Anyway, I'm keeping your suggestion in mind and I'll see if solves issues when we'll have to add mode resources or mode boards/soc derivates based on this porting.

galak commented:
> From what I see, ARM musca defines 2 boards, so does cypress_062_wifi_bt dual core board, let put these aside.
> I see your point about consistency. But then, looking at lpc5Xxxx implementations, I don't particularly agree on the way things are done (specially in soc files Kconfig.defconfig.lpcxxxx_mx some kind of peripheral assignment is done, which limits the flexibility at board level and performs driver deactivation at soc level which seems error prone for an end user).
> Anyway, I'm keeping your suggestion in mind and I'll see if solves issues when we'll have to add mode resources or mode boards/soc derivates based on this porting.

Musca is 2 different actual boards.  The cypress_062_wifi_bt should be reworked but we've never gotten around to it.
